src/shared/__tests__/start-gateway.test.js

```diff
@@ -276,7 +276,157 @@ describe("#start-gateway", () => {
+            jest.spyOn(process, "memoryUsage").mockReturnValue({
+                rss: 1025 * 1024 * 1024,
+            });
+            await startGateway(options, pretendApp);
+            const infoSpy = jest.spyOn(options.logger, "info");
+            const useCallback = useMock.mock.calls[0][0];
+
+            // Act
+            useCallback();
+
+            // Assert
+            expect(infoSpy).toHaveBeenCalledWith(
+                "Memory usage has gone over maximum. (used: 1074790400), limit: 864026624",
+            );
+            expect(infoSpy).not.toHaveBeenCalledWith(
+                "Gracefully shutting down server.",
+            );
+        });
+
+        it("should attempt to shutdown the server", async () => {
+            // Arrange
+            process.env.GAE_MEMORY_MB = "1024";
+            process.env.MIN_FREE_MB = "200";
+            const options = {
+                name: "TEST_GATEWAY",
+                port: 42,
+                host: "127.0.0.1",
+                logger: createLogger("test", "debug"),
+                mode: "test",
+            };
+            const fakeServer = {
+                address: () => ({
+                    address: "ADDRESS",
+                    port: "PORT",
+                }),
+                close: jest.fn(),
+            };
+            const listenMock = jest.fn().mockReturnValue(fakeServer);
+            const useMock = jest.fn().mockReturnValue(null);
+            const pretendApp = ({
+                listen: listenMock,
+                use: useMock,
+            }: any);
+            jest.spyOn(
+                UseAppEngineMiddleware,
+                "useAppEngineMiddleware",
+            ).mockReturnValue(Promise.resolve(pretendApp));
+            jest.spyOn(process, "on").mockReturnValue(null);
+            jest.spyOn(process, "memoryUsage").mockReturnValue({
+                rss: 1025 * 1024 * 1024,
+            });
+            await startGateway(options, pretendApp);
+            const infoSpy = jest.spyOn(options.logger, "info");
+            const useCallback = useMock.mock.calls[0][0];
+
+            // Act
+            useCallback();
+
+            // Assert
+            expect(infoSpy).toHaveBeenCalledWith(
+                "Memory usage has gone over maximum. (used: 1074790400), limit: 864026624",
+            );
+            expect(infoSpy).toHaveBeenCalledWith(
+                "Gracefully shutting down server.",
+            );
+            expect(fakeServer.close).toHaveBeenCalled();
+        });
+    });
+
     describe("close on SIGINT", () => {
+        const GAE_MEMORY_MB = process.env.GAE_MEMORY_MB;
+        const MIN_FREE_MB = process.env.MIN_FREE_MB;
+
+        beforeEach(() => {
+            delete process.env.GAE_MEMORY_MB;
+            delete process.env.MIN_FREE_MB;
+        });
+
+        afterEach(() => {
+            process.env.GAE_MEMORY_MB = GAE_MEMORY_MB;
+            process.env.MIN_FREE_MB = MIN_FREE_MB;
```

Comment:
This won't work right if the envvars were absent. Instead, it needs to be something like:

```
if (GAE_MEMORY_MB == null) {
    delete process.env.GAE_MEMORY_MB;
} else {
    process.env.GAE_MEMORY_MB = GAE_MEMORY_MB;
}
```

I wrote a helper for this somewhere, but that might've been in some webapp tests.